[MODULE2-PART-5] Add ability to signup for users.

snikolenko-flo opened the pull request:
Features

- Signup new users
- Filter images by user
- Hash users' passwords 
- Use passport.js middleware to login users

snikolenko-flo committed:
feature(gallery): add a sign up endpoint

add a sign up endpoint

MODULE-2-PART-4-mongodb

snikolenko-flo committed:
feature(gallery): login user via Passport.js local strategy

login user via Passport.js local strategy

MODULE-2-PART-5-auth

snikolenko-flo committed:
feature(gallery): create an authorization middleware for a login validation

create an authorization middleware for a login validation

MODULE-2-PART-5-auth

snikolenko-flo committed:
feature(gallery): connect users and images

connect users and images db collections

MODULE-2-PART-5-auth

snikolenko-flo committed:
feature(gallery): add user filter to query string

add user filter to query string

MODULE-2-PART-5-auth

snikolenko-flo committed:
refactor(gallery): optimize bd query for user images

optimzie bd query for user images

MODULE-2-PART-5-auth

snikolenko-flo committed:
chore(gallery): update readme according to the latest changes

update readme according to the latest changes

MODULE-2-PART-5-auth

arevta-flo commented on frontend/src/signup/handler.ts:
```diff
@@ -0,0 +1,26 @@
+import { SignupManager } from './signup.manager.js';
```
signup duplicates the code of login.

arevta-flo commented on frontend/src/html/signup.html:
```diff
@@ -0,0 +1,18 @@
+<!DOCTYPE html>
```
I think we need to make it possible to move to this page from login page

arevta-flo commented on backend/src/models/user.model.ts:
```diff
@@ -1,8 +1,27 @@
 import mongoose, {Schema} from 'mongoose';
+import bcrypt from 'bcrypt';
```
we can avoid this dependency and use built in nodejs modules

arevta-flo commented on backend/src/models/user.model.ts:
```diff
@@ -1,8 +1,27 @@
+UserSchema.pre(
+  'save',
+  async function(next) {
+    const user = this;
```
I think we can just use 'this'

arevta-flo commented on backend/src/services/db-service.ts:
```diff
@@ -114,6 +118,13 @@ export class DbService {
   }
 
+  async getUserImagesNumber(user): Promise<number> {
+    const userImages = await User.find({'email': user}).select(['images']);
```
it is the same code as on 159 line.  we can move it into separate function.

arevta-flo commented on frontend/src/gallery/gallery.api.ts:
```diff
@@ -2,9 +2,14 @@ import { BASE_URL } from '../data/constants.js';
     const accessToken = localStorage.getItem('token');
-    const url = `${BASE_URL}/gallery?page=${pageNumber}&limit=${pageLimit}`;
+
+    let url = `${BASE_URL}/gallery?page=${pageNumber}&limit=${pageLimit}`;
```
you can refactor it like this:
let url =  `${BASE_URL}/gallery?page=${pageNumber}`
url = user ? `${url}&filter=${user}` : url;

arevta-flo commented on frontend/src/gallery/gallery.url.ts:
```diff
@@ -5,8 +5,11 @@ export class GalleryUrl {
-  addParametersToUrl(pageNumber: number, pageLimit: number): void {
-    const urlInAddressBar = `./gallery.html?page=${pageNumber}&limit=${pageLimit}`;
+  addParametersToUrl(pageNumber: number, pageLimit: number, user?: string): void {
+    let urlInAddressBar = `./gallery.html?page=${pageNumber}&limit=${pageLimit}`;
```
the same refactoring as in fetchImages function

arevta-flo commented on backend/src/models/user.model.ts:
```diff
@@ -2,8 +2,8 @@ import mongoose, {Schema} from 'mongoose';
-  password: {type: String, required: true},
+  email: { type: String, required: true, unique: true },
+  password: { type: String, required: true },
   images: [{ type: Schema.Types.ObjectId, ref: 'Image' }]
```
you've already connected each image to appropriate users in images schema. i think that you can avoid referencing to image schema inside user schema.

snikolenko-flo committed:
refactor(gallery): remove reference to images from a user schema

remove reference to images from a user schema

MODULE-2-PART-5-auth

snikolenko-flo committed:
refactor(gallery): create separate function for getting user images

create separate function for getting user images

MODULE-2-PART-5-auth

snikolenko-flo committed:
refactor(gallery): remove if statement from url building

remove if statement from url building

MODULE-2-PART-5-auth

snikolenko-flo committed:
refactor(gallery): use built-in node.js module instead of bcrypt

use built-in node.js module instead of bcrypt

MODULE-2-PART-5-auth

snikolenko-flo committed:
feature(gallery): add a link to the signup page

add a link to the signup page

MODULE-2-PART-5-auth

snikolenko-flo committed:
refactor(gallery): get rid of code duplication in the signup module

get rid of code duplication in the signup module

MODULE-2-PART-5-auth

snikolenko-flo committed:
refactor(gallery): install nodemon as dev dependency

install nodemon as dev dependency

MODULE-2-PART-5-auth

snikolenko-flo committed:
refactor(gallery): put jwt secret to the env file

put jwt secret to the env file

MODULE-2-PART-5-auth

arevta-flo commented on frontend/src/gallery/gallery.api.ts:
```diff
@@ -6,10 +6,7 @@ export class GalleryApi {
-    if(user) {
-      url = `${BASE_URL}/gallery?page=${pageNumber}&limit=${pageLimit}&filter=${user}`;
-    }
+    url = user ? `${BASE_URL}/gallery?page=${pageNumber}&limit=${pageLimit}&filter=${user}` : url;
```
you don't need to duplicate url again)

arevta-flo commented on frontend/src/gallery/gallery.url.ts:
```diff
@@ -7,9 +7,8 @@ export class GalleryUrl {
-    if(user) {
-      urlInAddressBar = `./gallery.html?page=${pageNumber}&limit=${pageLimit}&filter=${user}`;
-    }
+    urlInAddressBar = user ? `./gallery.html?page=${pageNumber}&limit=${pageLimit}&filter=${user}` : urlInAddressBar;
```
the same here

arevta-flo commented on backend/src/models/user.model.ts:
```diff
@@ -1,27 +1,36 @@
 
 UserSchema.pre(
   'save',
   async function(next) {
```
it is async function but you use sync hashing function

arevta-flo commented on backend/src/models/user.model.ts:
```diff
@@ -1,27 +1,36 @@
   }
 );
 
 UserSchema.methods.isValidPassword = async function(password) {
```
the same here

snikolenko-flo committed:
refactor(gallery): use async version of the hash function

use async version of the hash function

MODULE-2-PART-5-auth

snikolenko-flo committed:
refactor(gallery): remove url duplication

remove url duplication

MODULE-2-PART-5-auth

snikolenko-flo committed:
fix(gallery): redirect user to the correct page after user filter request

redirect user to the correct page after user filter request

MODULE-2-PART-5-auth

snikolenko-flo committed:
feature(gallery): add whitelist with allowed origins

add whitelist with allowed origins

MODULE-2-PART-5-auth

snikolenko-flo committed:
feature(gallery): configure cors without cors npm

configure cors without cors npm

MODULE-2-PART-5-auth
